adapt/engine.py

```diff
@@ -175,6 +175,62 @@ def register_intent_parser(self, intent_parser):
         else:
             raise ValueError("%s is not an intent parser" % str(intent_parser))
 
+    def drop_intent_parser(self, parser_name):
+        """Drop a registered intent parser.
+
+        Arguments:
+            parser_name (str): parser name to drop.
+
+        Returns:
+            (bool) True if a parser was dropped else False
+        """
+        new_parsers = [p for p in self.intent_parsers if p.name != parser_name]
+        num_original_parsers = len(self.intent_parsers)
+        self.intent_parsers = new_parsers
+
+        return len(self.intent_parsers != num_original_parsers)
+
+    def drop_entity(self, pattern):
+        """Drop all entities mathching the given pattern
+
+        Arguments:
+            pattern: regex pattern to match against
+        """
+        removables = []
+        regex = re.compile(pattern)
+
+        def _find_type(node, current=''):
+            """Search Trie for nodes with a matching beginning.
+
+            Performs depth first search of the Trie removing the references.
+
+            Arguments:
+                node (trie node): Node to parse
+                current (str): string "position" in Trie
+            """
+            nonlocal removables
+            nonlocal regex
+            # Check if node matches
+            vocab_types = [d[1] for d in node.data if len(d[1])]
+            matches = any([regex.match(v) for v in vocab_types])
+
+            if matches:
+                if len(vocab_types) == 1:
+                    # If this is the last mark it as a removable node
+                    removables.append(current)
+                else:
+                    # just remove the matching type from data list
+                    node.data = [d for d in node.data
+                                 if not regex.match(d[1])]
+
+            for c in node.children:
+                _find_type(node.children[c], current + c)
+
+        _find_type(self.trie.root)
+        # Remove any nodes that would be completely empty
+        for word in removables:
+            self.trie.remove(word)
```

Comment:
I think there's a gap here where two skills may have registered the same vocab word. There are tests for trie.remove that assert we can simple_remove (like here), and named_remove (which we use for inserting aliases), but nothing that covers a simple remove when an alias/collision exists within the trie. 

Imagine the following:
```python3
trie.insert('foo', 'FOO') # inserted by SkillA
trie.insert('foo', 'BAR') # inserted by SkillB

trie.remove('foo')
```

I'm not actually sure what the outcome would be, and we don't have a test to cover it.

Reply:
I'll write some test cases but the code above should not add it to removables if there are more intents pointing to it

Reply:
Got it, I think I've got a slightly clearer understanding now; I apologize, this is some of the most-opaque code, and I regret not having any structure/typing to the objects stored in the Trie. After a little more digging, I have a couple concerns. 

1) This implementation breaks the interface of the Trie; it's not expected that an external caller is going to be traversing the structure itself. In an ideal world, I would have labeled `Trie.root` as `Trie._root` to communicate that expectation of privacy, but alas, I was so young!

2) I believe that the current implementation of `Trie.remove` can actually capture most of this, depending on some expectations of the caller, and the value of pattern. If you can reconstruct the tuple that gets stuffed into the `Trie` by `IntentDeterminationEngine`, then you should be able to simply use `trie.remove(value, data=(...))`. I don't know enough about what the caller is going to do at this point, but that seems like a reasonable input. There's a small memory leak where, when the `data` param is provided, `trie.remove` will not delete empty nodes. 

3) `Trie.insert` currently de-dups on collisions, so if two skills add the same vocab (which is valid), and one deregisters, it could break the other.

4) When unloading a skill, this would not appear to support `regex_entitiy` cleanup.

Reply:
Thanks for the feedback. Gonna dig into the Trie again and see if I can get a better understanding of it.

3. This is sort of handled in Mycroft, skills all register with unique keywords (skill creators Keywords gets a skill-id prepended to it.) Hence the regex search so a `my_skill_id.*' can be used to find all of a skills registered keywords.

4. I was leaving that out for a future PR since my brain never wants to remember how regex actually works in adapt

So maybe we could do the following, to not break the interface of the Trie

- Add a `reverse_search` method to the Trie (using a matching function as an argument as data object is not really known to the Trie itself) to find all nodes id's and data matching.
- Take the results using Trie.remove

Reply:
> Add a reverse_search method to the Trie (using a matching function as an argument as data object is not really known to the Trie itself) to find all nodes id's and data matching.

Yup, this is exactly what I was thinking! Let me know if you need any help.

Reply:
I'll push something tonight along with some testcases